File: apps/sdkwork-im-h5/packages/sdkwork-im-h5-chat/src/services/offlineSendQueue.ts
import { readImH5IamSessionTokens } from "@sdkwork/im-h5-core";

const LEGACY_SESSION_STORAGE_KEY = "sdkwork-im-h5:pending-sends:v1";
const IDB_NAME = "sdkwork-im-h5-offline";
const IDB_VERSION = 1;
const IDB_STORE = "pending_sends";
const DEFAULT_FLUSH_LIMIT = 50;
const MAX_PENDING_SENDS = 100;

export interface PendingTextSendPayload {
  conversationId: string;
  text: string;
  clientMsgId: string;
}

interface PendingSendRecord {
  tenantId: string;
  clientMsgId: string;
  conversationId: string;
  payloadJson: string;
  createdAt: string;
  attemptCount: number;
  flushClaimId?: string | null;
}

type PendingSendPayloadWithClaim = PendingTextSendPayload & {
  clientMsgId: string;
  claimId: string;
};

let idbReady: Promise<IDBDatabase | null> | null = null;
let pendingSendFlushInFlight: Promise<void> | null = null;

function resolveTenantId(): string | undefined {
  const context = readImH5IamSessionTokens()?.context;
  const tenantId = context?.tenantId?.trim();
  return tenantId || undefined;
}

function createPendingSendClaimId(): string {
  return `h5-flush-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`;
}

function openOfflineDatabase(): Promise<IDBDatabase | null> {
  if (typeof indexedDB === "undefined") {
    return Promise.resolve(null);
  }
  if (!idbReady) {
    idbReady = new Promise((resolve) => {
      const request = indexedDB.open(IDB_NAME, IDB_VERSION);
      request.onupgradeneeded = () => {
        const db = request.result;
        if (!db.objectStoreNames.contains(IDB_STORE)) {
          const store = db.createObjectStore(IDB_STORE, {
            keyPath: ["tenantId", "clientMsgId"],
          });
          store.createIndex("tenant_created", ["tenantId", "createdAt"], { unique: false });
          store.createIndex("tenant_claim", ["tenantId", "flushClaimId"], { unique: false });
        }
      };
      request.onsuccess = () => {
        void migrateLegacySessionStorage(request.result).finally(() => {
          resolve(request.result);
        });
      };
      request.onerror = () => resolve(null);
    });
  }
  return idbReady;
}

async function migrateLegacySessionStorage(db: IDBDatabase): Promise<void> {
  if (typeof window === "undefined") {
    return;
  }
  const raw = window.sessionStorage.getItem(LEGACY_SESSION_STORAGE_KEY);
  if (!raw) {
    return;
  }
  try {
    const parsed: unknown = JSON.parse(raw);
    if (!Array.isArray(parsed)) {
      return;
    }
    for (const item of parsed) {
      if (!item || typeof item !== "object" || Array.isArray(item)) {
        continue;
      }
      const record = item as PendingSendRecord;
      if (
        typeof record.tenantId !== "string"
        || typeof record.clientMsgId !== "string"
        || typeof record.conversationId !== "string"
        || typeof record.payloadJson !== "string"
      ) {
        continue;
      }
      await putRecord(db, {
        tenantId: record.tenantId,
        clientMsgId: record.clientMsgId,
        conversationId: record.conversationId,
        payloadJson: record.payloadJson,
        createdAt: typeof record.createdAt === "string" ? record.createdAt : new Date().toISOString(),
        attemptCount: typeof record.attemptCount === "number" ? record.attemptCount : 0,
        flushClaimId: null,
      });
    }
  } catch {
    // Drop corrupt legacy queue.
  } finally {
    window.sessionStorage.removeItem(LEGACY_SESSION_STORAGE_KEY);
  }
}

function runReadWrite<T>(db: IDBDatabase, operation: (store: IDBObjectStore) => void): Promise<T> {
  return new Promise((resolve, reject) => {
    const transaction = db.transaction(IDB_STORE, "readwrite");
    const store = transaction.objectStore(IDB_STORE);
    let result: T;
    try {
      result = operation(store) as T;
    } catch (error) {
      transaction.abort();
      reject(error);
      return;
    }
    transaction.oncomplete = () => resolve(result);
    transaction.onerror = () => reject(transaction.error ?? new Error("idb transaction failed"));
    transaction.onabort = () => reject(transaction.error ?? new Error("idb transaction aborted"));
  });
}

function runReadonly<T>(db: IDBDatabase, operation: (store: IDBObjectStore) => IDBRequest<T>): Promise<T> {
  return new Promise((resolve, reject) => {
    const transaction = db.transaction(IDB_STORE, "readonly");
    const store = transaction.objectStore(IDB_STORE);
    const request = operation(store);
    request.onsuccess = () => resolve(request.result);
    request.onerror = () => reject(request.error ?? new Error("idb read failed"));
  });
}

function collectIndexRecords(db: IDBDatabase, indexName: string, key: IDBValidKey): Promise<PendingSendRecord[]> {
  return new Promise((resolve, reject) => {
    const transaction = db.transaction(IDB_STORE, "readonly");
    const index = transaction.objectStore(IDB_STORE).index(indexName);
    const records: PendingSendRecord[] = [];
    const request = index.openCursor(IDBKeyRange.only(key));
    request.onsuccess = () => {
      const cursor = request.result;
      if (!cursor) {
        resolve(records);
        return;
      }
      records.push(cursor.value as PendingSendRecord);
      cursor.continue();
    };
    request.onerror = () => reject(request.error ?? new Error("idb cursor failed"));
  });
}

function putRecord(db: IDBDatabase, record: PendingSendRecord): Promise<void> {
  return runReadWrite(db, (store) => {
    store.put(record);
  });
}

function deleteRecord(db: IDBDatabase, tenantId: string, clientMsgId: string): Promise<void> {
  return runReadWrite(db, (store) => {
    store.delete([tenantId, clientMsgId]);
  });
}

async function withDatabase<T>(operation: (db: IDBDatabase) => Promise<T>): Promise<T | undefined> {
  const db = await openOfflineDatabase();
  if (!db) {
    return undefined;
  }
  return operation(db);
}

function parsePayload(record: PendingSendRecord): PendingTextSendPayload | undefined {
  try {
    const parsed: unknown = JSON.parse(record.payloadJson);
    if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) {
      return undefined;
    }
    const payload = parsed as PendingTextSendPayload;
    if (
      typeof payload.conversationId !== "string"
      || typeof payload.text !== "string"
      || typeof payload.clientMsgId !== "string"
    ) {
      return undefined;
    }
    return payload;
  } catch {
    return undefined;
  }
}

export function isRetryableH5SendError(error: unknown): boolean {
  if (error instanceof TypeError) {
    return true;
  }
  const message = error instanceof Error ? error.message : String(error);
  const normalized = message.toLowerCase();
  return (
    normalized.includes("failed to fetch")
    || normalized.includes("network")
    || normalized.includes("timeout")
    || normalized.includes("service unavailable")
    || normalized.includes("503")
    || normalized.includes("502")
    || normalized.includes("504")
  );
}

export async function enqueuePendingTextSend(payload: PendingTextSendPayload): Promise<void> {
  const tenantId = resolveTenantId();
  if (!tenantId) {
    return;
  }
  await withDatabase(async (db) => {
    const tenantRecords = (await collectIndexRecords(db, "tenant_created", tenantId))
      .filter((record) => !record.flushClaimId);
    if (tenantRecords.length >= MAX_PENDING_SENDS) {
      const sorted = [...tenantRecords].sort((left, right) => left.createdAt.localeCompare(right.createdAt));
      const dropCount = tenantRecords.length - MAX_PENDING_SENDS + 1;
      for (const record of sorted.slice(0, dropCount)) {
        await deleteRecord(db, record.tenantId, record.clientMsgId);
      }
    }
    await putRecord(db, {
      tenantId,
      clientMsgId: payload.clientMsgId,
      conversationId: payload.conversationId,
      payloadJson: JSON.stringify(payload),
      createdAt: new Date().toISOString(),
      attemptCount: 0,
      flushClaimId: null,
    });
  });
}

export async function listPendingTextSends(
  limit = DEFAULT_FLUSH_LIMIT,
): Promise<PendingTextSendPayload[]> {
  const tenantId = resolveTenantId();
  if (!tenantId) {
    return [];
  }
  const rows = await withDatabase(async (db) => {
    const records = (await collectIndexRecords(db, "tenant_created", tenantId))
      .filter((record) => !record.flushClaimId)
      .sort((left, right) => left.createdAt.localeCompare(right.createdAt))
      .slice(0, limit);
    return records;
  });
  if (!rows) {
    return [];
  }
  const payloads: PendingTextSendPayload[] = [];
  for (const record of rows) {
    const payload = parsePayload(record);
    if (payload) {
      payloads.push(payload);
    }
  }
  return payloads;
}

export async function claimPendingTextSends(
  limit = DEFAULT_FLUSH_LIMIT,
): Promise<PendingSendPayloadWithClaim[]> {
  const tenantId = resolveTenantId();
  if (!tenantId) {
    return [];
  }
  const claimId = createPendingSendClaimId();
  const rows = await withDatabase(async (db) => {
    const candidates = (await collectIndexRecords(db, "tenant_created", tenantId))
      .filter((record) => !record.flushClaimId)
      .sort((left, right) => left.createdAt.localeCompare(right.createdAt))
      .slice(0, limit);
    for (const record of candidates) {
      await putRecord(db, {
        ...record,
        flushClaimId: claimId,
        attemptCount: record.attemptCount + 1,
      });
    }
    return collectIndexRecords(db, "tenant_claim", [tenantId, claimId]);
  });
  if (!rows) {
    return [];
  }
  const payloads: PendingSendPayloadWithClaim[] = [];
  for (const record of rows.sort((left, right) => left.createdAt.localeCompare(right.createdAt))) {
    const payload = parsePayload(record);
    if (payload) {
      payloads.push({ ...payload, claimId });
    }
  }
  return payloads;
}

export async function releasePendingTextSendClaim(
  clientMsgId: string,
  claimId: string,
): Promise<void> {
  const tenantId = resolveTenantId();
  if (!tenantId) {
    return;
  }
  await withDatabase(async (db) => {
    const record = await runReadonly<PendingSendRecord | undefined>(db, (store) => store.get([tenantId, clientMsgId]));
    if (!record || record.flushClaimId !== claimId) {
      return;
    }
    await putRecord(db, { ...record, flushClaimId: null });
  });
}

export async function removePendingTextSend(clientMsgId: string): Promise<void> {
  const tenantId = resolveTenantId();
  if (!tenantId) {
    return;
  }
  await withDatabase(async (db) => {
    await deleteRecord(db, tenantId, clientMsgId);
  });
}

export async function runPendingTextSendFlush(
  flush: (pending: PendingSendPayloadWithClaim[]) => Promise<void>,
): Promise<void> {
  if (pendingSendFlushInFlight) {
    await pendingSendFlushInFlight;
    return;
  }
  pendingSendFlushInFlight = (async () => {
    const pending = await claimPendingTextSends();
    if (pending.length === 0) {
      return;
    }
    await flush(pending);
  })().finally(() => {
    pendingSendFlushInFlight = null;
  });
  await pendingSendFlushInFlight;
}

File: apps/sdkwork-im-flutter-mobile/packages/sdkwork_im_flutter_mobile_chat/lib/src/services/offline_send_queue.dart
import 'dart:convert';
import 'dart:io';

import 'package:shared_preferences/shared_preferences.dart';

const _storageKey = 'sdkwork-im-flutter-mobile:pending-sends:v2';
const _legacyStorageKey = 'sdkwork-im-flutter-mobile:pending-sends:v1';
const _defaultFlushLimit = 50;
const maxPendingSends = 100;

class PendingTextSendPayload {
  const PendingTextSendPayload({
    required this.conversationId,
    required this.text,
    required this.clientMsgId,
  });

  final String conversationId;
  final String text;
  final String clientMsgId;

  Map<String, dynamic> toJson() => {
        'conversationId': conversationId,
        'text': text,
        'clientMsgId': clientMsgId,
      };

  factory PendingTextSendPayload.fromJson(Map<String, dynamic> json) {
    return PendingTextSendPayload(
      conversationId: json['conversationId']?.toString() ?? '',
      text: json['text']?.toString() ?? '',
      clientMsgId: json['clientMsgId']?.toString() ?? '',
    );
  }
}

class PendingTextSendPayloadWithClaim extends PendingTextSendPayload {
  const PendingTextSendPayloadWithClaim({
    required super.conversationId,
    required super.text,
    required super.clientMsgId,
    required this.claimId,
  });

  final String claimId;
}

class _PendingSendRecord {
  const _PendingSendRecord({
    required this.tenantId,
    required this.clientMsgId,
    required this.conversationId,
    required this.payloadJson,
    required this.createdAt,
    this.attemptCount = 0,
    this.flushClaimId,
  });

  final String tenantId;
  final String clientMsgId;
  final String conversationId;
  final String payloadJson;
  final String createdAt;
  final int attemptCount;
  final String? flushClaimId;

  Map<String, dynamic> toJson() => {
        'tenantId': tenantId,
        'clientMsgId': clientMsgId,
        'conversationId': conversationId,
        'payloadJson': payloadJson,
        'createdAt': createdAt,
        'attemptCount': attemptCount,
        if (flushClaimId != null) 'flushClaimId': flushClaimId,
      };

  factory _PendingSendRecord.fromJson(Map<String, dynamic> json) {
    return _PendingSendRecord(
      tenantId: json['tenantId']?.toString() ?? '',
      clientMsgId: json['clientMsgId']?.toString() ?? '',
      conversationId: json['conversationId']?.toString() ?? '',
      payloadJson: json['payloadJson']?.toString() ?? '',
      createdAt: json['createdAt']?.toString() ?? '',
      attemptCount: int.tryParse(json['attemptCount']?.toString() ?? '') ?? 0,
      flushClaimId: json['flushClaimId']?.toString(),
    );
  }

  _PendingSendRecord copyWith({
    int? attemptCount,
    String? flushClaimId,
    bool clearFlushClaimId = false,
  }) {
    return _PendingSendRecord(
      tenantId: tenantId,
      clientMsgId: clientMsgId,
      conversationId: conversationId,
      payloadJson: payloadJson,
      createdAt: createdAt,
      attemptCount: attemptCount ?? this.attemptCount,
      flushClaimId: clearFlushClaimId ? null : (flushClaimId ?? this.flushClaimId),
    );
  }
}

Future<void>? _pendingSendFlushInFlight;

String _createPendingSendClaimId() {
  return 'flutter-flush-${DateTime.now().millisecondsSinceEpoch}-${DateTime.now().microsecond}';
}

bool isRetryableFlutterSendError(Object error) {
  if (error is SocketException || error is HttpException) {
    return true;
  }
  final message = error.toString().toLowerCase();
  return message.contains('failed host lookup')
      || message.contains('connection refused')
      || message.contains('connection reset')
      || message.contains('network')
      || message.contains('timeout')
      || message.contains('service unavailable')
      || message.contains('503')
      || message.contains('502')
      || message.contains('504');
}

Future<List<_PendingSendRecord>> _readQueue(SharedPreferences prefs) async {
  final raw = prefs.getString(_storageKey);
  if (raw == null || raw.isEmpty) {
    return const [];
  }
  try {
    final decoded = jsonDecode(raw);
    if (decoded is! List) {
      return const [];
    }
    return decoded
        .whereType<Map>()
        .map((entry) => _PendingSendRecord.fromJson(Map<String, dynamic>.from(entry)))
        .toList();
  } catch (_) {
    return const [];
  }
}

Future<void> _writeQueue(SharedPreferences prefs, List<_PendingSendRecord> records) async {
  await prefs.setString(
    _storageKey,
    jsonEncode(records.map((record) => record.toJson()).toList()),
  );
}

Future<void> _migrateLegacyQueue(SharedPreferences prefs) async {
  final legacyRaw = prefs.getString(_legacyStorageKey);
  if (legacyRaw == null || legacyRaw.isEmpty) {
    return;
  }
  try {
    final decoded = jsonDecode(legacyRaw);
    if (decoded is! List) {
      return;
    }
    final migrated = <_PendingSendRecord>[];
    for (final entry in decoded.whereType<Map>()) {
      final record = _PendingSendRecord.fromJson(Map<String, dynamic>.from(entry));
      if (record.tenantId.isEmpty || record.clientMsgId.isEmpty) {
        continue;
      }
      migrated.add(
        record.copyWith(clearFlushClaimId: true),
      );
    }
    if (migrated.isNotEmpty) {
      await _writeQueue(prefs, migrated);
    }
  } catch (_) {
    // Drop corrupt legacy queue.
  } finally {
    await prefs.remove(_legacyStorageKey);
  }
}

Future<SharedPreferences> _openPrefs() async {
  final prefs = await SharedPreferences.getInstance();
  await _migrateLegacyQueue(prefs);
  return prefs;
}

PendingTextSendPayload? _parsePayload(_PendingSendRecord record) {
  try {
    final decoded = jsonDecode(record.payloadJson);
    if (decoded is! Map) {
      return null;
    }
    final payload = PendingTextSendPayload.fromJson(
      Map<String, dynamic>.from(decoded),
    );
    if (payload.conversationId.isEmpty
        || payload.text.isEmpty
        || payload.clientMsgId.isEmpty) {
      return null;
    }
    return payload;
  } catch (_) {
    return null;
  }
}

Future<void> enqueuePendingTextSend({
  required String tenantId,
  required PendingTextSendPayload payload,
}) async {
  if (tenantId.isEmpty) {
    return;
  }
  final prefs = await _openPrefs();
  final queue = (await _readQueue(prefs))
      .where((record) => record.clientMsgId != payload.clientMsgId)
      .toList();
  queue.add(
    _PendingSendRecord(
      tenantId: tenantId,
      clientMsgId: payload.clientMsgId,
      conversationId: payload.conversationId,
      payloadJson: jsonEncode(payload.toJson()),
      createdAt: DateTime.now().toUtc().toIso8601String(),
      attemptCount: 0,
    ),
  );
  final tenantQueue = queue.where((record) => record.tenantId == tenantId).toList();
  if (tenantQueue.length > maxPendingSends) {
    final sorted = [...tenantQueue]
      ..sort((left, right) => left.createdAt.compareTo(right.createdAt));
    final dropCount = tenantQueue.length - maxPendingSends;
    final dropIds = sorted
        .take(dropCount)
        .map((record) => record.clientMsgId)
        .toSet();
    queue.removeWhere(
      (record) => record.tenantId == tenantId && dropIds.contains(record.clientMsgId),
    );
  }
  await _writeQueue(prefs, queue);
}

Future<List<PendingTextSendPayload>> listPendingTextSends({
  required String tenantId,
  int limit = _defaultFlushLimit,
}) async {
  if (tenantId.isEmpty) {
    return const [];
  }
  final prefs = await _openPrefs();
  final payloads = <PendingTextSendPayload>[];
  final records = (await _readQueue(prefs))
      .where((record) => record.tenantId == tenantId && record.flushClaimId == null)
      .toList()
    ..sort((left, right) => left.createdAt.compareTo(right.createdAt));
  for (final record in records.take(limit)) {
    final payload = _parsePayload(record);
    if (payload != null) {
      payloads.add(payload);
    }
  }
  return payloads;
}

Future<List<PendingTextSendPayloadWithClaim>> claimPendingTextSends({
  required String tenantId,
  int limit = _defaultFlushLimit,
}) async {
  if (tenantId.isEmpty) {
    return const [];
  }
  final prefs = await _openPrefs();
  final claimId = _createPendingSendClaimId();
  final queue = await _readQueue(prefs);
  final candidates = queue
      .where((record) => record.tenantId == tenantId && record.flushClaimId == null)
      .toList()
    ..sort((left, right) => left.createdAt.compareTo(right.createdAt));
  final selected = candidates.take(limit).map((record) => record.clientMsgId).toSet();
  if (selected.isEmpty) {
    return const [];
  }
  final updated = queue
      .map((record) {
        if (record.tenantId == tenantId && selected.contains(record.clientMsgId)) {
          return record.copyWith(
            flushClaimId: claimId,
            attemptCount: record.attemptCount + 1,
          );
        }
        return record;
      })
      .toList();
  await _writeQueue(prefs, updated);
  final payloads = <PendingTextSendPayloadWithClaim>[];
  for (final record in updated.where(
    (record) => record.tenantId == tenantId && record.flushClaimId == claimId,
  )) {
    final payload = _parsePayload(record);
    if (payload != null) {
      payloads.add(
        PendingTextSendPayloadWithClaim(
          conversationId: payload.conversationId,
          text: payload.text,
          clientMsgId: payload.clientMsgId,
          claimId: claimId,
        ),
      );
    }
  }
  payloads.sort((left, right) => left.clientMsgId.compareTo(right.clientMsgId));
  return payloads;
}

Future<void> releasePendingTextSendClaim({
  required String tenantId,
  required String clientMsgId,
  required String claimId,
}) async {
  if (tenantId.isEmpty || claimId.isEmpty) {
    return;
  }
  final prefs = await _openPrefs();
  final queue = await _readQueue(prefs);
  final updated = queue
      .map((record) {
        if (record.tenantId == tenantId
            && record.clientMsgId == clientMsgId
            && record.flushClaimId == claimId) {
          return record.copyWith(clearFlushClaimId: true);
        }
        return record;
      })
      .toList();
  await _writeQueue(prefs, updated);
}

Future<void> removePendingTextSend({
  required String tenantId,
  required String clientMsgId,
}) async {
  if (tenantId.isEmpty) {
    return;
  }
  final prefs = await _openPrefs();
  final queue = (await _readQueue(prefs))
      .where(
        (record) => !(record.tenantId == tenantId && record.clientMsgId == clientMsgId),
      )
      .toList();
  await _writeQueue(prefs, queue);
}

Future<void> runPendingTextSendFlushForTenant({
  required String tenantId,
  required Future<void> Function(List<PendingTextSendPayloadWithClaim> pending) flush,
  int limit = _defaultFlushLimit,
}) async {
  if (_pendingSendFlushInFlight != null) {
    await _pendingSendFlushInFlight;
    return;
  }
  _pendingSendFlushInFlight = () async {
    final pending = await claimPendingTextSends(tenantId: tenantId, limit: limit);
    if (pending.isEmpty) {
      return;
    }
    await flush(pending);
  }();
  try {
    await _pendingSendFlushInFlight;
  } finally {
    _pendingSendFlushInFlight = null;
  }
}
